R/eml_data_qa.R

```diff
@@ -60,53 +48,79 @@ qa_package <- function(node, pid, readData = TRUE) {
+        cat(crayon::green(paste0("\n\n..................Checking access, metadata..................")))
+        sysmeta <- dataone::getSystemMetadata(node, package$metadata)
+        qa_access(sysmeta, creator_ORCIDs)
+
+        # Check resource_map
+        cat(crayon::green(paste0("\n\n..................Checking access, resource_map..................")))
+        sysmeta <- dataone::getSystemMetadata(node, package$resource_map)
+        qa_access(sysmeta, creator_ORCIDs)
+    }
 
-    for (i in seq_along(eml@dataset@dataTable)) {
-        dataTable <- eml@dataset@dataTable[[i]]
+    urls_dataTable <- unique(unlist(EML::eml_get(eml@dataset@dataTable, "url"), recursive = T))
+    urls_otherEntity <- unique(unlist(EML::eml_get(eml@dataset@otherEntity, "url"), recursive = T))
+    urls <- unique(append(urls_dataTable, urls_otherEntity))
 
-        if (length(dataTable@physical) == 0) {
-            cat(crayon::red(paste0("\nMissing physical for dataTable with entityName: ", dataTable@entityName)))
-            next
-        }
+    # Checks that each data object has a matching url in the eml.
+    wrong_URL <- FALSE
+    for (i in seq_along(package$data)) {
+        data <- package$data[i] # use this as opposed to (data in package$data) to preserve names
+        n <- which(grepl(paste0(data, "$"), urls))
 
-        if (length(dataTable@physical[[1]]@distribution) == 0) {
-            cat(crayon::red(paste0("\nMissing URL/distribution info for dataTable with objectName: ", dataTable@physical[[1]]@objectName)))
+        if (length(n) == 1) {
+            cat(crayon::green(paste0("\nThe URL/distribution for ", names(data), " is correctly listed in the physical section of the eml.")))
         } else {
-            urls <- c(urls, dataTable@physical[[1]]@distribution[[1]]@online@url)
+            cat(crayon::red(paste0("\nThe URL/distribution for ", names(data), " is missing/incongruent in the physical section of the eml.")))
+            wrong_URL <- TRUE
         }
     }
 
-    if (length(urls) == 0) {
-        cat(crayon::red("\nNone of the objects in this package have associated physicals with URLs. Fix this error and try again."))
-        return(0)
+    if (length(urls) != length(package$data) | wrong_URL) {
+        # Must stop here to ensure proper ordering in following loops.
+        stop("\nAll distribution URLs in the eml must match the package's data URLs to continue.",
+             "\nPlease fix and try again.")
     }
 
     for (objectpid in package$data)
     {
-        # Does the given pid have an associated datatable
-        sysmeta <- dataone::getSystemMetadata(node, objectpid)
 
-        # If object is not tabular data, continue
-        format <- sysmeta@formatId
-        if (!format %in% supported_file_formats) next
-
-        # If object is tabular data, should also have a dataTable, which is matched by the url
-        i <- which(grepl(objectpid, urls))
+        n_dT <- which(grepl(paste0(objectpid, "$"), urls_dataTable))
```

Comment:
I think the ordering of these operations should be changed to the following -
- Get sysmeta
- Check access (if flag is true)
- Check if file is of supported types, if not, next
- Then, check for matching pids in urls
We don't need to check for datatables/other entities if the file type wouldn't have attributes, right?